src/Serilog/Parameters/PropertyBinder.cs

```diff
@@ -51,45 +51,62 @@ public IEnumerable<LogEventProperty> ConstructProperties(MessageTemplate message
         }
 
-        IEnumerable<LogEventProperty> ConstructPositionalProperties(MessageTemplate template, object[] messageTemplateParameters)
+        IEnumerable<LogEventProperty> ConstructPositionalPropertiesWithSomeMissing(MessageTemplate template, object[] messageTemplateParameters)
         {
             var positionalProperties = template.PositionalProperties;
 
-            if (positionalProperties.Length != messageTemplateParameters.Length)
-                SelfLog.WriteLine("Positional property count does not match parameter count: {0}", template);
-
             var result = new LogEventProperty[messageTemplateParameters.Length];
-            foreach (var property in positionalProperties)
+            for (int position = 0; position < messageTemplateParameters.Length; position++)
             {
-                int position;
-                if (property.TryGetPositionalValue(out position))
+                PropertyToken propertyToken = null;
+                for (int propIndex = 0; propIndex < positionalProperties.Length; propIndex++)
                 {
-                    if (position < 0 || position >= messageTemplateParameters.Length)
-                        SelfLog.WriteLine("Unassigned positional value {0} in: {1}", position, template);
-                    else
-                        result[position] = ConstructProperty(property, messageTemplateParameters[position]);
+                    int propertyTokenPosition;
+                    if (positionalProperties[propIndex].TryGetPositionalValue(out propertyTokenPosition)
+                        && propertyTokenPosition == position)
+                    {
+                        propertyToken = positionalProperties[propIndex];
+                    }
                 }
+
+                result[position] = propertyToken == null
+                    ? new LogEventProperty("__" + position,
+                        _valueConverter.CreatePropertyValue(messageTemplateParameters[position]))
+                    : ConstructProperty(propertyToken, messageTemplateParameters[position]);
             }
 
-            var next = 0;
-            for (var i = 0; i < result.Length; ++i)
+            return result;
+        }
+
+        IEnumerable<LogEventProperty> ConstructPositionalPropertiesOptimisedAllMatching(MessageTemplate template, object[] messageTemplateParameters)
+        {
+            var positionalProperties = template.PositionalProperties;
+            var result = new LogEventProperty[positionalProperties.Length];
+            for (var index = 0; index < positionalProperties.Length; index++)
             {
-                if (result[i] != null)
+                var property = positionalProperties[index];
+                int position;
+                if (!property.TryGetPositionalValue(out position))
                 {
-                    result[next] = result[i];
-                    ++next;
+                    throw new Exception("should never happen here");
```

Comment:
How aboue `new ArgumentException("ConstructPositionalPropertiesOptimisedAllMatching() may only be called when template and provided parameters match exactly.")` ?